### DISCUSSION TITLE
[help]Why can't I use "serial.close()" to close the occupancy of the Serial port?

### DISCUSSION THREAD

**Discussion by Conmi-WhiteJoker** [Pi4J/pi4j#103] [Q&A]
When the program exits, I use "Serial.close ()".Serial is NULL, but I am still using it until I exit without clearing it.
I found that the serial port was still occupied when I restarted the application.
`java.io.IOException: Failed to SET terminal attributes for the serial port. (Error #22)`

**Comment by Conmi-WhiteJoker** [comment]
`@PreDestroy 
private void closeUart() throws Exception {
    if (!serial.isClosed()) {
        serial.close(); 
    }
}`

**Comment by luixxx** [comment]
Hi Conmi-WhiteJoker, 
I can't help, :( 
i'm newbie in Pi4j-V2, could you share an example for use of Serial ports?
Thankyou

**Reply by FDelporte** [reply]
Example has been created on https://pi4j.com/documentation/io-examples/serial/

**Comment by savageautomate** [comment]
@Conmi-WhiteJoker 

The `Serial` instance is `Auto-Closable` and should perform a `close()` on it's own when the Pi4J context is `shutdown()`.

See:
https://github.com/Pi4J/pi4j-v2/blob/32efbf668a9a49fa8e53c55ef23e765180a49aa1/pi4j-core/src/main/java/com/pi4j/io/serial/Serial.java#L39

Also if you are using the PiGPIO library for Serial communication, it appears to also try and ensure the serial is closed on a shutdown. 

https://github.com/Pi4J/pi4j-v2/blob/32efbf668a9a49fa8e53c55ef23e765180a49aa1/libraries/pi4j-library-pigpio/src/main/java/com/pi4j/library/pigpio/impl/PiGpioNativeImpl.java#L118-L123

So if the Serial instance is not shutting down properly that would be a problem.  
Are you explicitly calling the `shutdown()` method of the Pi4J context object?

https://github.com/Pi4J/pi4j-example-minimal/blob/5afb12d0388c246f67d5cda8c98a21960f7d2847/src/main/java/com/pi4j/example/MinimalExample.java#L157-L158
